sql/partitioning/partition_base.cc

```diff
@@ -0,0 +1,5823 @@
+                                                  char *      table_name_buf,
+                                                  size_t table_name_buf_len)
+{
+  DBUG_ASSERT(canonical_name);
+  DBUG_ASSERT(db_name_buf);
+  DBUG_ASSERT(db_name_buf_len);
+  DBUG_ASSERT(db_name_buf_len);
+  DBUG_ASSERT(table_name_buf);
+
+  const char *db_name_begin= strchr(canonical_name, FN_LIBCHAR);
+  const char *db_name_end= strrchr(canonical_name, FN_LIBCHAR);
+
+  DBUG_ASSERT(db_name_begin);
+  DBUG_ASSERT(db_name_end);
+  DBUG_ASSERT(db_name_begin != db_name_end);
+
+  ++db_name_begin;
+  size_t db_name_size= db_name_end - db_name_begin;
+
+  DBUG_ASSERT(db_name_size + 1 < db_name_buf_len);
+
+  memcpy(db_name_buf, db_name_begin, db_name_size);
+  db_name_buf[db_name_size]= '\0';
+
+  const char *table_name= db_name_end + 1;
+  DBUG_ASSERT(*table_name);
+  size_t table_name_size= strlen(table_name);
+  DBUG_ASSERT(table_name_size + 1 < table_name_buf_len);
+
+  memcpy(table_name_buf, table_name, table_name_size);
+  table_name_buf[table_name_size]= '\0';
+}
+
+bool get_part_str_for_path(const char *path, std::string &result)
+{
+  // Check for .frm file existence
+  MY_STAT frm_stat_info;
+  if (!my_stat(path, &frm_stat_info, MYF(0)))
+  {
+    // if .frm file does not exist don't treat it as error
+    if (my_errno() == ENOENT)
+      return true;
+    return false;
+  }
+
+  // First, we open the file, and return upon failure. No need to close
+  // the file in this case.
+  File file= mysql_file_open(key_file_frm, path, O_RDONLY | O_SHARE, MYF(0));
+  if (file < 0)
+    return false;
+  BOOST_SCOPE_EXIT_ALL(&) { mysql_file_close(file, MYF(MY_WME)); };
+
+  // Next, we read the header and do some basic verification of the
+  // header fields.
+  uchar head[64];
+  if (mysql_file_read(file, head, sizeof(head), MYF(MY_NABP)))
+    return false;
+
+  if (!strncmp((char *)head, "TYPE=VIEW\n", strlen("TYPE=VIEW\n")))
+    return true;
+
+  if (head[0] != (uchar)254 || head[1] != 1 ||
+      !(head[2] == FRM_VER || head[2] == FRM_VER + 1 ||
+        (head[2] >= FRM_VER + 3 && head[2] <= FRM_VER + 4)))
+    // Upon failure, return NULL, but here, we have to close the file first.
+    return false;
+
+  // For other engines, and for cluster tables with version >= 50120, we
+  // continue by checking that we have an extra data segment and a proper
+  // form position.
+  const ulong pos= get_form_pos(file, head);   //< Position of form info
+  const uint  n_length= uint4korr(head + 55);  //< Length of extra segment
+  if (n_length == 0 || pos == 0)
+  {
+    // We close the file and return success, as we no form info
+    // or extra segment.
+    return false;
+  }
+
+  mysql_file_seek(file, pos, MY_SEEK_SET, MYF(0));
+
+  uchar *extra_segment_buff= static_cast<uchar *>(
+      my_malloc(key_memory_frm_extra_segment_buff, n_length, MYF(MY_WME)));
+  std::unique_ptr<uchar, decltype(&my_free)> extra_segment_buff_uptr(
```

Comment:
I would not do this, the pattern works well with standard library. This might make sense in 8.0 to reuse the code, but there is no reason to bring new dependency in 5.7.

Reply:
OK